scanner/families/plugins/runner/runner.go

```diff
@@ -67,77 +69,113 @@ func (s *Scanner) Scan(ctx context.Context, inputType common.InputType, userInpu
-	go func() {
-		logger := logger.WithField("metadata", map[string]interface{}{
-			"name":       to.ValueOrZero(metadata.Name),
-			"version":    to.ValueOrZero(metadata.Version),
-			"apiVersion": to.ValueOrZero(metadata.ApiVersion),
-		}).WithField("plugin", s.config.Name)
+		res := result{
+			Result: types.ScannerResult{},
+			Err:    nil,
+		}
 
-		logs, err := rr.Logs(ctx)
-		if err != nil {
-			logger.WithError(err).Warnf("could not listen for logs on plugin runner")
+		if err := rr.Start(ctx); err != nil {
+			res.Err = fmt.Errorf("failed to start plugin runner: %w", err)
+			resChan <- res
 			return
 		}
-		defer logs.Close()
 
-		for r := bufio.NewScanner(logs); r.Scan(); {
-			logger.Info(r.Text())
+		if err := rr.WaitReady(ctx); err != nil {
+			res.Err = fmt.Errorf("failed to wait for plugin scanner to be ready: %w", err)
+			resChan <- res
+			return
 		}
-	}()
 
-	if err := rr.Run(ctx); err != nil {
-		finishRunner(ctx)
-		return nil, fmt.Errorf("failed to run plugin scanner: %w", err)
-	}
+		// Get plugin metadata
+		metadata, err := rr.Metadata(ctx)
+		if err != nil {
+			res.Err = fmt.Errorf("failed to get plugin scanner metadata: %w", err)
+			resChan <- res
+			return
+		}
 
-	if err := rr.WaitDone(ctx); err != nil {
-		finishRunner(ctx)
-		return nil, fmt.Errorf("failed to wait for plugin scanner to finish: %w", err)
-	}
+		// Stream logs
+		go func() {
+			logger := logger.WithField("metadata", map[string]interface{}{
+				"name":       to.ValueOrZero(metadata.Name),
+				"version":    to.ValueOrZero(metadata.Version),
+				"apiVersion": to.ValueOrZero(metadata.ApiVersion),
+			}).WithField("plugin", s.config.Name)
+
+			logs, err := rr.Logs(ctx)
+			if err != nil {
+				logger.WithError(err).Warnf("could not listen for logs on plugin runner")
+				return
+			}
+			defer logs.Close()
+
+			for r := bufio.NewScanner(logs); r.Scan(); {
+				logger.Info(r.Text())
+			}
+		}()
+
+		if err := rr.Run(ctx); err != nil {
+			res.Err = fmt.Errorf("failed to run plugin scanner: %w", err)
+			resChan <- res
+			return
+		}
 
-	findings, pluginResult, err := s.parseResults(ctx, rr)
-	if err != nil {
-		finishRunner(ctx)
-		return nil, fmt.Errorf("failed to parse plugin scanner results: %w", err)
-	}
+		if err := rr.WaitDone(ctx); err != nil {
+			res.Err = fmt.Errorf("failed to wait for plugin scanner to finish: %w", err)
+			resChan <- res
+			return
+		}
 
-	finishRunner(ctx)
+		findings, pluginResult, err := s.parseResults(ctx, rr)
+		if err != nil {
+			res.Err = fmt.Errorf("failed to parse plugin scanner results: %w", err)
+			resChan <- res
+			return
+		}
 
-	return &types.ScannerResult{
-		Findings: findings,
-		Output:   pluginResult,
-	}, nil
+		res.Result.Findings = findings
+		res.Result.Output = pluginResult
+		resChan <- res
+	}(ctx)
+
+	select {
+	case <-ctx.Done():
+		shutdownRunner(ctx)
+		return nil, errors.New("plugin context cancelled")
+	case r := <-resChan:
+		shutdownRunner(ctx)
+		return &types.ScannerResult{
```

Comment:
we can receive error here but it is not handled, we always assume the received result is valid

Reply:
Not exactly sure that I understand. Should I return an empty `ScannerResult` struct here?

`return &types.ScannerResult{}, errors.New("plugin context cancelled")`

Reply:
resChan result contains error which is not checked

Reply:
Ah, thanks! Fixed.